File: C++/Regular ExpressionMatching.cpp
#include<iostream>
#include<vector>
#include<queue>
#include<string>
#include<string.h>
#include<math.h>
#include<algorithm>
using namespace std;
class Solution {
public:
	bool match(const char *s, int sindex, int sizes, const char *p, int pindex, int sizep)
	{
		if(sindex > sizes)
			return false;
		if(sindex == sizes && pindex == sizep)
			return true;
		
		if(pindex < sizep-1 && p[pindex+1] == '*')
		{
			if(p[pindex] == '.')
			{
				for(int i = sindex; i <= sizes; i++)
				{
					if(match(s, i, sizes, p, pindex+2, sizep))
						return true;
				}
				return false;
			}
			else
			{
				int i = sindex;
				for(; i <= sizes; i++)
				{
					if(s[i] != p[pindex])
						break;
					if(match(s, i, sizes, p, pindex+2, sizep))
						return true;
				}
				return match(s, i ,sizes, p, pindex+2, sizep);
			}
		}
		else if(p[pindex] == '.')
			return match(s, sindex+1, sizes, p, pindex+1, sizep);
		else
		{
			if(s[sindex] == p[pindex])
				return match(s, sindex+1, sizes, p, pindex+1, sizep);
			else
				return false;
		}
	}
	bool isMatch(const char *s, const char *p) {
		// Start typing your C/C++ solution below
		// DO NOT write int main() function    
		if(!s && !p)
			return true;
		else if(!s || !p)
			return false;
		int lens = strlen(s);
		int lenp = strlen(p);
		return match(s, 0, lens, p,0, lenp);
	}
};
int main()
{
	char s[] = "";
	char p[] = ".ac*";
	Solution sol;
	cout<<sol.isMatch(s,p);
	getchar();
}

File: JAVA/RegularExpressionMatching.java
class Solution {
    public boolean isMatch(String s, String p) {
        // Start typing your Java solution below
        // DO NOT write main() function
        if(s == null && p == null)
        	return true;
        else if(s == null || p == null)
        	return false;
        return isMatch(s, 0, p, 0);
    }

	private boolean isMatch(String s, int ss, String p, int ps) {
		if(ps >= p.length())
			return ss >= s.length();
			
		if(ps+1 == p.length())
		{
			if(ss+1 == s.length() && (s.charAt(ss) == p.charAt(ps) || p.charAt(ps) == '.'))
				return true;
			return false;
		}
		if(p.charAt(ps+1) != '*')
		{
			if(ss < s.length() && (s.charAt(ss) == p.charAt(ps) || p.charAt(ps) == '.'))
				return isMatch(s, ss+1, p, ps+1);
			return false;
		}
		else
		{
			while(ss < s.length() && (s.charAt(ss) == p.charAt(ps) || p.charAt(ps) == '.'))
			{
				if(isMatch(s, ss, p, ps+2))
					return true;
				ss++;
			}
			return isMatch(s, ss, p, ps+2);
			
		}
	}
}
public class RegularExpressionMatching {

	/**
	 * @param args
	 */
	public static void main(String[] args) {
		// TODO Auto-generated method stub

	}

}
